packages/adapters-library/src/adapters/pendle/products/market/pendleMarketAdapter.ts

```diff
@@ -0,0 +1,305 @@
+
+import { Helpers } from '../../../../scripts/helpers'
+import { IProtocolAdapter } from '../../../../types/IProtocolAdapter'
+import {
+  AssetType,
+  GetEventsInput,
+  GetPositionsInput,
+  GetTotalValueLockedInput,
+  MovementsByBlock,
+  PositionType,
+  ProtocolAdapterParams,
+  ProtocolDetails,
+  ProtocolPosition,
+  ProtocolTokenTvl,
+  TokenType,
+  UnwrapExchangeRate,
+  UnwrapInput,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { Protocol } from '../../../protocols'
+import { fetchAllMarkets } from '../../backend/backendSdk'
+
+import { logger } from '../../../../core/utils/logger'
+import { OraclePyYtLp__factory } from '../../contracts'
+
+type Metadata = Record<
+  string,
+  {
+    protocolToken: Erc20Metadata
+    underlyingToken: Erc20Metadata
+    marketAddress: string
+    type: 'pt' | 'yt' | 'lp' | 'sy'
+  }
+>
+
+export class PendleMarketAdapter implements IProtocolAdapter, IMetadataBuilder {
+  productId = 'market'
+  protocolId: Protocol
+  chainId: Chain
+  helpers: Helpers
+
+  private provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+    helpers,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+  }
+
+  async unwrap({
+    blockNumber,
+    protocolTokenAddress,
+    tokenId,
+  }: UnwrapInput): Promise<UnwrapExchangeRate> {
+    const metadata = (await this.buildMetadata())[protocolTokenAddress]
+    const underlyingToken = metadata!.underlyingToken
+
+    const oracle = OraclePyYtLp__factory.connect(
+      '0x9a9fa8338dd5e5b2188006f1cd2ef26d921650c2',
+      this.provider,
+    )
+
+    let rate: bigint
+    switch (metadata!.type) {
+      case 'pt':
+        rate = await oracle
+          .getPtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'yt':
+        rate = await oracle
+          .getYtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'lp':
+        rate = await oracle
+          .getPtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'sy': {
+        rate = 10n ** 18n
+
```

Comment:
Doesn't need the second `!`, and the first one can be avoided to if it's fetch with `fetchPoolMetadata`.